### PR TITLE
fixed bug some times return 3 element after $this->redis->exec()

### PR DESCRIPTION
**Pull request by davidtall**
see #64

### CODE REVIEW

File: src/Resque/Worker.php [hunk 1]
```diff
@@ -760,9 +760,9 @@ public function queueDelayed($endTime = null, $startTime = 0)
             $this->redis->multi();
             $jobs = $this->redis->zrangebyscore(Queue::redisKey($queue, 'delayed'), $startTime, $endTime);
             $this->redis->zremrangebyscore(Queue::redisKey($queue, 'delayed'), $startTime, $endTime);
-            $res = $this->redis->exec();
-            if(!is_array($res) || count($res) != 2) return;
-            list($jobs, $found) = $res ;
+            $execReturn = $this->redis->exec();
+            if(!is_array($execReturn) || count($execReturn) != 2) return;
```
Comment (xelan) [inline src/Resque/Worker.php]: According to PSR-2, this should be
```php
if (!is_array($execReturn) || count($execReturn) !== 2) {
    return;
}
```
Note: The comparision is changed to a type-safe one, as count() must return an integer according to the [docs](http://php.net/count).

File: src/Resque/Worker.php [hunk 2]
```diff
@@ -760,9 +760,9 @@ public function queueDelayed($endTime = null, $startTime = 0)
...
+            list($jobs, $found) = $execReturn ;
```
Comment (xelan) [inline src/Resque/Worker.php]: Here should be no space before the semicolon.

**Comment by francislavoie** [comment]
To me this honestly feels like a bit of a bandaid because we don't have a proper understanding of why the bug even happens. I don't think just doing a `return;` is a safe plan because then we can't really predict what will happen. I'm not super comfortable with this change as it stands.

Did you read the discussion in the issue @xelan? Do you have any ideas?

**Comment by xelan** [comment]
Thank you, @davidtall and @francislavoie! I agree that finding the root cause is the better option. At the moment, I set up a reproducable test setup for the bug. I'll try to use the branch of this PR to see if it helps for now, then we can investigate into the reasons of the bug.

**Comment by xelan** [comment]
Apparently, this issue has to do with [Redis transaction handling](https://redis.io/topics/transactions). In [Worker.php](https://github.com/mjphaynes/php-resque/blob/master/src/Resque/Worker.php#L759), the transaction is started with `$this->redis->multi()`, then `$this->redis->zremrangebyscore() is called two times .` After that, the transaction is executed and the results are assigned.

The problem is probably that another worker may issue a command at the same time, and as the Redis documentation states, transactions are handled per client. If two workers are running in parallel with the same connection, the isolation probably can't be guaranteed.

According to a quick grep through our codebase, `Worker.php` is the only place where a transaction is initiated via the `multi()` method.

**Comment by xelan** [comment]
In the same constellation (two workers on the same machine, same Redis connection, one is cancelled), another error (PHP notice) can appear:
```
** Done job shop252:My\Job\ExampleJob#991f7a3715e8fd3fc6d567() in 1.2ms
^CPHP Notice:  Object of class Predis\Response\Status could not be converted to int in /path/to/worker/vendor/mjphaynes/php-resque/src/Resque/Worker.php on line 765
PHP Notice:  Object of class Predis\Response\Status could not be converted to int in /path/to/worker/vendor/mjphaynes/php-resque/src/Resque/Helpers/Stats.php on line 59
PHP Notice:  Object of class Predis\Response\Status could not be converted to int in /path/to/worker/vendor/mjphaynes/php-resque/src/Resque/Helpers/Stats.php on line 59
PHP Notice:  Object of class Predis\Response\Status could not be converted to int in /path/to/worker/vendor/mjphaynes/php-resque/src/Resque/Worker.php on line 780
```

**Comment by francislavoie** [comment]
@davidtall so, were you forking the worker process to do your scaling? The recommended way to run this is by running each worker as separate processes. If you do that, then they should each have unique Redis connections and errors like this wouldn't happen.

**Comment by xelan** [comment]
I've now published 2.1.2 with the bugfix for processing the last queue. Thanks again to @davidtall and @francislavoie for your great work on that!

For this issue/PR, I would investigate a bit more before releasing a so-so bugfix. 

What happens if we just return, as proposed in this PR? Are the delayed jobs just processed one interval cycle later or is there a danger of data loss? In my opinion, if another operation somehow gets into the transaction, the exec would cause all three to be commited, so the job deletion (`zremrangebyscore`) would already take place. If we just return after that, the values stored in Redis (queued jobs) are lost.

The problem is that we can't get rid of the transaction here. If a new job gets enqueued just between the `zrangebyscore()` and `zremrangebyscore()` calls, the latter one would remove it although it has never been processed. So we need atomic handling using a Redis transaction in this case.

Maybe a better solution would be to check the returned values of the commit in more detail to find out which and how many jobs were returned - and just ignore any other data `exec()` might return. 

Of course, that approach would also be a kind of "workaround". In the end, workers are supposed to run as separate processes with distinct Redis connections, avoiding the problem at its root.

@mjphaynes What do you think?

**Comment by davidtall** [comment]
@xelan  your mean is every worker should run with distinct Redis DB? And one woker use one Redis DB？So，I should push jobs into different Redis DB

**Comment by xelan** [comment]
The same Redis database, but every worker should be independent of the others, for example run in a different container. That way, workers can be scaled easily and problems with one worker do not influence the others.

I can post a small Docker example when I have time.

**Comment by davidtall** [comment]
@xelan Oh，I got it, your mean is workers should run in different machine or different container , 
 for example, using Docker

**Comment by xelan** [comment]
Yes, or at least separate PHP processes to avoid that the same Redis connection is reused. The best practice is to use one container (e.g. Docker) per worker.

**Comment by xelan** [comment]
A documentation about worker best practices will follow, so I'm closing this PR. Thanks again for the contribution and discussion :smile:

**Comment by scones** [comment]
Technically one could run multiple workers in one process, but this would require to fork the workers out of the main process and connect them to redis separately.

This is not supported by any resque implementation tho, as it is simpler and robust to start several workers, either by hand or docker-compose scale, or docker swarm scale or even kubernetes autoscaling.
The main concern with this is downscaling, as the wrong signals might kill a running child job, leaving the  data in an `undefined` state.

**Pull request closed without merging**